File: web/src/lib/decision.ts
/** The team's which-analysis rule — THE single web-side source of truth.
 *
 *  MIRROR: ratings_module_build_kit/decision.py is the same rule for the sync worker.
 *  Edit both together; tests on both sides pin the boundary values.
 *
 *  Rule v1 (validated on 8 months of ratings data, Sep 2026 — see the Rating-Threshold study):
 *  escalation → video, always · rating ≥ 4.55 → none · < 5 ratings → watch ·
 *  ≥ 40% of attendees rated → video · under 40% → transcript.
 *
 *  Rule v2 (Sep 2026 — the Instructor-Approval study; analysis/approval_rule.py is the spec):
 *  two bars decide IF a class enters the queue — rated below 4.55, or fewer than 80% of the
 *  room would have the instructor back (either one, with at least 5 voices) — and the weighted
 *  Class Health Score (rating 60 / approval 30 / track record 10, each 100 at its bar and 0 at
 *  its floor) decides HOW URGENT and HOW DEEP: urgent (< 70) → video whatever the reach ·
 *  borderline (≥ 90) → transcript first · in between, the 40% reach bar decides as in v1.
 *  A missing vote or a missing track record scores 100 — a missing signal is never a penalty.
 */

export const GOOD = 4.55;
export const MIN_VOICES = 5;
export const PARTICIPATION_BAR = 40;

export const LINE = GOOD;
export const APPROVAL_BAR = 80;
export const REACH_BAR = PARTICIPATION_BAR;
export const R_FLOOR = 3.55;
export const A_FLOOR = 40;
export const T_FLOOR = 4.05;
export const T_MIN_CLASSES = 3;
export const WEIGHTS = { rating: 0.6, approval: 0.25, track: 0.15 } as const;
export const URGENT = 70;
export const BORDERLINE = 90;

export type Decision = "none" | "watch" | "transcript" | "video";
export type HealthBand = "urgent" | "look" | "borderline";
export type FlagReason = "rating" | "approval" | "escalated";

export const BAND_LABEL: Record<HealthBand, string> = {
  urgent: "Urgent",
  look: "Needs a look",
  borderline: "Borderline",
};

export function decide(
  rating: number | null | undefined,
  numRatings: number | null | undefined,
  attended: number | null | undefined,
  escalated = false,
): Decision {
  if (escalated) return "video";
  if (rating == null || Number.isNaN(rating)) return "watch";
  if (rating >= GOOD) return "none";
  if (numRatings == null || Number.isNaN(numRatings) || numRatings < MIN_VOICES) return "watch";
  if (!attended) return "watch";
  return (numRatings / attended) * 100 >= PARTICIPATION_BAR ? "video" : "transcript";
}

export function participationPct(
  numRatings: number | null | undefined,
  attended: number | null | undefined,
): number | null {
  if (numRatings == null || !attended) return null;
  return Math.round((numRatings / attended) * 100);
}

/** One-line human explanation per verdict (the queue chips and the form helper share these). */
export function explain(d: Decision, pct: number | null, numRatings: number | null): string {
  switch (d) {
    case "none":
      return `Rating is ${GOOD} or above — no analysis needed unless a PM asks.`;
    case "watch":
      return numRatings != null && numRatings < MIN_VOICES
        ? `Only ${numRatings} learner${numRatings === 1 ? "" : "s"} rated it — one or two opinions, not a class problem yet. Watch the next session.`
        : "Not enough data to judge yet — watch.";
    case "video":
      return `Below ${GOOD} and ${pct}% of attendees rated it (≥ ${PARTICIPATION_BAR}%) — a representative share of the room spoke and it was still low.`;
    case "transcript":
      return `Below ${GOOD} but only ${pct}% of attendees rated it (< ${PARTICIPATION_BAR}%) — too thin to trust yet; a transcript read is enough.`;
  }
}

// ── rule v2 ──────────────────────────────────────────────────────────────────
const clamp = (v: number, lo = 0, hi = 100) => Math.max(lo, Math.min(hi, v));

/** 0 at R_FLOOR, 100 at the line. */
export const scoreRating = (rating: number) => clamp((100 * (rating - R_FLOOR)) / (LINE - R_FLOOR));

/** 0 at A_FLOOR, 100 at the bar. No vote → 100 (no penalty). */
export const scoreApproval = (approval: number | null | undefined) =>
  approval == null ? 100 : clamp((100 * (approval - A_FLOOR)) / (APPROVAL_BAR - A_FLOOR));

/** 0 at T_FLOOR, 100 at the line. No track record → 100 (no penalty). */
export const scoreTrack = (trackAvg: number | null | undefined) =>
  trackAvg == null ? 100 : clamp((100 * (trackAvg - T_FLOOR)) / (LINE - T_FLOOR));

/** The Class Health Score, 0–100, unrounded. The band reads THIS value; the verdict and the DB
 *  carry it rounded to one decimal — so a stored 90.0 can sit in "look" (it was 89.99). */
export function healthScore(
  rating: number,
  approval: number | null | undefined,
  trackAvg: number | null | undefined,
): number {
  return (
    WEIGHTS.rating * scoreRating(rating) +
    WEIGHTS.approval * scoreApproval(approval) +
    WEIGHTS.track * scoreTrack(trackAvg)
  );
}

export function healthBand(score: number): HealthBand {
  return score < URGENT ? "urgent" : score < BORDERLINE ? "look" : "borderline";
}

/** yes/(yes+no) as a percent; null when either count is unknown or nobody voted. */
export function approvalPct(
  yesVotes: number | null | undefined,
  noVotes: number | null | undefined,
): number | null {
  if (yesVotes == null || noVotes == null || Number.isNaN(yesVotes) || Number.isNaN(noVotes)) return null;
  const votes = yesVotes + noVotes;
  return votes > 0 ? Math.round((yesVotes / votes) * 10000) / 100 : null;
}

/** "13 of 15 · 87%" — the vote as the queue prints it; null when nobody voted. */
export function voteLabel(
  yesVotes: number | null | undefined,
  noVotes: number | null | undefined,
): string | null {
  const pct = approvalPct(yesVotes, noVotes);
  if (pct == null) return null;
  return `${yesVotes} of ${yesVotes! + noVotes!} · ${Math.round(pct)}%`;
}

export type DecideV2Input = {
  rating: number | null | undefined;
  numRatings: number | null | undefined;
  attended: number | null | undefined;
  yesVotes?: number | null;
  noVotes?: number | null;
  trackAvg?: number | null;
  escalated?: boolean;
};

export type DecisionV2 = {
  decision: Decision;
  healthScore: number | null;
  healthBand: HealthBand | null;
  flagReasons: FlagReason[];
};

const num = (v: number | null | undefined) => (v == null || Number.isNaN(v) ? null : v);

export function decideV2(i: DecideV2Input): DecisionV2 {
  const rating = num(i.rating);
  const escalated = i.escalated ?? false;
  if (rating == null) {
    return {
      decision: escalated ? "video" : "watch",
      healthScore: null,
      healthBand: null,
      flagReasons: escalated ? ["escalated"] : [],
    };
  }
  const approval = approvalPct(i.yesVotes, i.noVotes);
  const flagReasons: FlagReason[] = [];
  if (rating < GOOD) flagReasons.push("rating");
  if (approval != null && approval < APPROVAL_BAR) flagReasons.push("approval");
  if (escalated) flagReasons.push("escalated");
  const raw = healthScore(rating, approval, i.trackAvg);
  const band = healthBand(raw);
  const out = (decision: Decision, healthBand: HealthBand | null): DecisionV2 => ({
    decision,
    healthScore: Math.round(raw * 10) / 10,
    healthBand,
    flagReasons,
  });
  if (escalated) return out("video", band);
  if (flagReasons.length === 0) return out("none", null);
  const voices = num(i.numRatings);
  if (voices == null || voices < MIN_VOICES) return out("watch", null);
  if (band === "urgent") return out("video", band);
  if (band === "borderline") return out("transcript", band);
  if (!i.attended) return out("watch", null);
  return out((voices / i.attended) * 100 >= PARTICIPATION_BAR ? "video" : "transcript", band);
}

/** One plain-English sentence for a v2 verdict, e.g.
 *  "Rated 4.31 and only 67% would have the instructor back — urgent, video." */
export function explainV2(v: DecisionV2, i: DecideV2Input): string {
  const rating = num(i.rating);
  if (i.escalated) return "Escalation reported — video, whatever the numbers say.";
  if (rating == null) return "No rating yet — watch.";
  const approval = approvalPct(i.yesVotes, i.noVotes);
  const back = approval == null ? null : `${Math.round(approval)}% would have the instructor back`;
  const r = rating.toFixed(2);
  const voices = num(i.numRatings);
  const reach = participationPct(voices, i.attended);
  const lowRating = v.flagReasons.includes("rating");
  const lowApproval = v.flagReasons.includes("approval");

  if (v.decision === "none") {
    return back
      ? `Rated ${r} and ${back} — clears both bars, no analysis unless a PM asks.`
      : `Rated ${r} with no vote recorded — above the line, no analysis unless a PM asks.`;
  }
  if (v.decision === "watch") {
    if (voices != null && voices < MIN_VOICES)
      return `Only ${voices} learner${voices === 1 ? "" : "s"} rated it — too thin for either signal; watch the next session.`;
    if (voices == null) return "Fill in how many learners rated it — fewer than 5 is too thin for either signal.";
    return "Attendance is missing, so the reach can't be judged — watch.";
  }
  const why =
    lowRating && lowApproval
      ? `Rated ${r} and only ${Math.round(approval!)}% would have the instructor back`
      : lowRating
        ? back
          ? `Rated ${r}, though ${back}`
          : `Rated ${r} with no vote recorded`
        : `Rated ${r} but only ${Math.round(approval!)}% would have the instructor back`;
  const how =
    v.healthBand === "urgent"
      ? "urgent, video"
      : v.healthBand === "borderline"
        ? "borderline, transcript first"
        : `needs a look, ${v.decision}${reach != null ? ` (${reach}% of the room rated it)` : ""}`;
  return `${why} — ${how}.`;
}

File: ratings_module_build_kit/decision.py
"""decision.py - LEGACY: the team's which-analysis rules v1 and v2, as pure functions.

Status (Feedback Loop v3, Sep 2026): the queue decision now comes from the Class Sentiment Score,
computed inside the database by score_class_rating() with the active scoring_configs row
(analysis/sentiment_score.py is the reference). The sync still writes the v2 read-out below
(approval_pct, track_avg, health_score, health_band, flag_reasons) for ONE release so the two
rules can be compared side by side, and the web mirror still explains it - nothing else reads it.
Do not extend this file; retire it with those columns.

Rule v1 (validated against 8 months of ratings data, Sep 2026 - the Rating-Threshold study) - decide():
  - any escalation                          -> video, always
  - rating 4.55 or above                    -> none (no analysis unless a PM asks)
  - fewer than 5 ratings                    -> watch (one or two opinions is not a class problem)
  - >= 40% of attendees rated it            -> video (representative sample, still bad)
  - under 40%                               -> transcript (too thin to trust yet)

Rule v2 (Sep 2026 - the Instructor-Approval study; analysis/approval_rule.py is the spec) - decide_v2():
two bars decide IF, the weighted Class Health Score decides HOW URGENT and HOW DEEP.
  - any escalation                                  -> video, always (reason 'escalated')
  - rating < 4.55  or  approval < 80%               -> the class has a problem (reasons 'rating' / 'approval')
        neither bar fails                           -> none
        fewer than 5 ratings                        -> watch (either signal is too thin)
  - Health = 0.60*R + 0.25*A + 0.15*T, each 0-100   -> band: urgent (<70) / look (70 to <90) / borderline (>=90)
    (weights W below - 60/25/15 from analysis/approval_weights.py; older notes saying 60/30/10 are stale)
  - depth: urgent -> video whatever the reach; borderline -> transcript whatever the reach;
           in between, >= 40% of attendees rated -> video, else transcript.
  An unknown approval (no vote) or an unknown track record (fewer than 3 earlier classes) scores
  100 - a missing signal is never a penalty.

Contract - what a Verdict carries (the DB columns and the web mirror use the same names):
  decision      'none' | 'watch' | 'transcript' | 'video'
  health_score  0-100 to one decimal; None only when the rating is unknown
  health_band   'urgent' | 'look' | 'borderline' when the class is queued (video/transcript), else None
                (read from the UNROUNDED score, as the study does - see health_score)
  flag_reasons  tuple drawn from ('rating', 'approval', 'escalated'), always in that order

MIRROR: web/src/lib/decision.ts is the same rule for the web app. Edit both together, and keep
the constants identical - the tests on both sides pin the boundary values.
"""
from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Optional

RULE_VERSION = "v2"

GOOD = 4.55          # at or above this line a class is fine (LINE in the study)
MIN_VOICES = 5       # fewer ratings than this and the score is one or two opinions
PARTICIPATION_BAR = 40.0   # percent of attendees who rated (REACH_BAR in the study)
APPROVAL_BAR = 80.0  # percent of voters who would have the instructor back

# Where each Health component hits 0 - "as bad as it gets" (analysis/approval_weights.py).
R_FLOOR = 3.55       # rating a full point under the line
A_FLOOR = 40.0       # approval: only 4 in 10 would have the instructor back
T_FLOOR = 4.05       # instructor's average half a point under the line
T_MIN_CLASSES = 3    # earlier classes needed before a track record counts

W = (0.60, 0.25, 0.15)  # weights: rating / approval / track record
URGENT = 70.0        # health below this -> 'urgent'
BORDERLINE = 90.0    # health at or above this -> 'borderline'; between -> 'look'

# Depth by score: urgent -> video, borderline -> transcript, whatever the reach. Switched off
# (DEPTH_BY_SCORE=0), depth is reach-only as in v1 - the bars and bands are unchanged.
DEPTH_BY_SCORE = (os.environ.get("DEPTH_BY_SCORE") or "1").strip().lower() not in ("0", "false", "no", "off")

DECISIONS = ("none", "watch", "transcript", "video")
BANDS = ("urgent", "look", "borderline")
REASONS = ("rating", "approval", "escalated")


def decide(rating, num_ratings, attended, escalated: bool = False) -> str:
    """One class's verdict under rule v1. Missing data degrades to 'watch', never to a confident verdict."""
    if escalated:
        return "video"
    if rating is None:
        return "watch"
    if float(rating) >= GOOD:
        return "none"
    if num_ratings is None or int(num_ratings) < MIN_VOICES:
        return "watch"
    if not attended:
        return "watch"                      # can't compute participation
    pct = int(num_ratings) / int(attended) * 100.0
    return "video" if pct >= PARTICIPATION_BAR else "transcript"


# ── rule v2 ──────────────────────────────────────────────────────────────────

def _clamp(v: float, lo: float = 0.0, hi: float = 100.0) -> float:
    return max(lo, min(hi, v))


def score_rating(rating) -> float:
    """0 at R_FLOOR, 100 at the line."""
    return _clamp(100.0 * (float(rating) - R_FLOOR) / (GOOD - R_FLOOR))


def score_approval(approval_pct) -> float:
    """0 at A_FLOOR, 100 at the bar. No vote -> 100 (no penalty)."""
    if approval_pct is None:
        return 100.0
    return _clamp(100.0 * (float(approval_pct) - A_FLOOR) / (APPROVAL_BAR - A_FLOOR))


def score_track(track_avg) -> float:
    """0 at T_FLOOR, 100 at the line. No track record -> 100 (no penalty)."""
    if track_avg is None:
        return 100.0
    return _clamp(100.0 * (float(track_avg) - T_FLOOR) / (GOOD - T_FLOOR))


def health_score(rating, approval_pct=None, track_avg=None) -> float:
    """The Class Health Score, 0-100, unrounded. The band reads THIS value; the Verdict and the
    DB carry it rounded to one decimal - so a stored 90.0 can sit in 'look' (it was 89.99)."""
    return (W[0] * score_rating(rating) + W[1] * score_approval(approval_pct)
            + W[2] * score_track(track_avg))


def health_band(score) -> str:
    s = float(score)
    return "urgent" if s < URGENT else ("look" if s < BORDERLINE else "borderline")


def approval_pct(yes_votes, no_votes) -> Optional[float]:
    """yes/(yes+no) as a percent to two decimals; None when either count is unknown or no one voted."""
    if yes_votes is None or no_votes is None:
        return None
    yes, votes = int(yes_votes), int(yes_votes) + int(no_votes)
    return round(yes / votes * 100.0, 2) if votes > 0 else None


@dataclass(frozen=True)
class Verdict:
    decision: str
    health_score: Optional[float]
    health_band: Optional[str]
    flag_reasons: tuple = ()


def decide_v2(rating, num_ratings, attended, escalated: bool = False,
              approval_pct: Optional[float] = None, track_avg: Optional[float] = None) -> Verdict:
    """One class's verdict under rule v2. Missing data degrades to 'watch', never to a confident verdict."""
    if rating is None:
        return Verdict("video" if escalated else "watch", None, None,
                       ("escalated",) if escalated else ())
    reasons = []
    if float(rating) < GOOD:
        reasons.append("rating")
    if approval_pct is not None and float(approval_pct) < APPROVAL_BAR:
        reasons.append("approval")
    if escalated:
        reasons.append("escalated")
    raw = health_score(rating, approval_pct, track_avg)
    score, band = round(raw, 1), health_band(raw)
    if escalated:
        return Verdict("video", score, band, tuple(reasons))
    if not reasons:
        return Verdict("none", score, None, ())
    if num_ratings is None or int(num_ratings) < MIN_VOICES:
        return Verdict("watch", score, None, tuple(reasons))
    if DEPTH_BY_SCORE and band == "urgent":
        return Verdict("video", score, band, tuple(reasons))
    if DEPTH_BY_SCORE and band == "borderline":
        return Verdict("transcript", score, band, tuple(reasons))
    if not attended:
        return Verdict("watch", score, None, tuple(reasons))   # can't compute participation
    pct = int(num_ratings) / int(attended) * 100.0
    return Verdict("video" if pct >= PARTICIPATION_BAR else "transcript", score, band, tuple(reasons))
